vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs

```diff
@@ -133,72 +128,66 @@ module internal CommonHelpers =
-                | _ -> 
-                    // Otherwise, we recompute
-                    let newData = scanSourceLine(sourceTokenizer, textLine, lineContents, lexState)
-                    sourceTextData.[i] <- Some newData
-                    newData
-                
-            lexState <- lineData.LexStateAtEndOfLine
-
-            if startLine <= i then
-                result.Add(lineData)
-
-            // If necessary, invalidate all subsequent lines after endLine
-            if endLine < lines.Count - 1 then 
-                match sourceTextData.[endLine+1] with 
-                | Some data  -> 
-                    if data.LexStateAtStartOfLine <> lexState then
-                        sourceTextData.ClearFrom (endLine+1)
-                | None -> ()
-
-        result
-
     let getColorizationData(documentKey: DocumentId, sourceText: SourceText, textSpan: TextSpan, fileName: string option, defines: string list, 
                             cancellationToken: CancellationToken) : List<ClassifiedSpan> =
-        try
-            let lines = sourceText.Lines
-            let startLine = lines.GetLineFromPosition(textSpan.Start).LineNumber
-            let endLine = lines.GetLineFromPosition(textSpan.End).LineNumber
-            
-            // Rescan the lines if necessary and report the information
-            let result = new List<ClassifiedSpan>()
-            for lineData in getSourceLineDatas(documentKey, sourceText, startLine, endLine, fileName, defines, cancellationToken) do
-                result.AddRange(lineData.ClassifiedSpans |> Seq.filter(fun token ->
-                    textSpan.Contains(token.TextSpan.Start) ||
-                    textSpan.Contains(token.TextSpan.End - 1) ||
-                    (token.TextSpan.Start <= textSpan.Start && textSpan.End <= token.TextSpan.End)))
-            result
-        with 
-        | :? System.OperationCanceledException -> reraise()
-        |  ex -> 
-            Assert.Exception(ex)
-            List<ClassifiedSpan>()
+         try
+             let sourceTokenizer = FSharpSourceTokenizer(defines, fileName)
+             let lines = sourceText.Lines
+             // We keep incremental data per-document.  When text changes we correlate text line-by-line (by hash codes of lines)
+             let sourceTextData = dataCache.GetValue(documentKey, fun key -> SourceTextData(lines.Count))
+ 
+             let startLine = lines.GetLineFromPosition(textSpan.Start).LineNumber
+             let endLine = lines.GetLineFromPosition(textSpan.End).LineNumber
+             // Go backwards to find the last cached scanned line that is valid
+             let scanStartLine = 
+                 let mutable i = startLine
+                 while i > 0 && (match sourceTextData.[i] with Some data -> not (data.IsValid(lines.[i])) | None -> true)  do
+                     i <- i - 1
+                 i
+             // Rescan the lines if necessary and report the information
```

Comment:
I'm 100% with you on this. It must be covered by tests, then rewritten in a more functional and safe manner. The number of things this algorithm operates is not as huge to make the code so imperative.